### PR TITLE
feat(ecs): ECS optimized Windows images

### LINKED ISSUE #2574: Support windows containers
With https://github.com/awslabs/aws-cdk/pull/2453, we are supporting other AMIs available through SSM parameters. We should also be able to support pulling windows containers:

https://docs.aws.amazon.com/AmazonECS/latest/developerguide/retrieve-ecs-optimized_AMI.html

### PR DESCRIPTION
**Pull request by nmussy**
* deprecate `EcsOptimizedAmi` for `EcsOptimizedImage`
* constructor(props) replaced by strongly typed static methods
* Windows AMI support
* deprecate `EcsOptimizedAmi`, `EcsOptimizedAmiProps`
* will require changes on v2 shipment #3398

Fixes #2574

----

Please read the [contribution guidelines](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md) and follow the pull-request checklist.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts [hunk 1]
```diff
@@ -242,6 +247,13 @@ export interface EcsOptimizedAmiProps {
    */
   readonly generation?: ec2.AmazonLinuxGeneration;
 
+  /**
+   * The Windows Server version to use.
+   *
+   * @default none, uses Linux generation
+   */
+  readonly windowsVersion?: WindowsOptimizedVersion;
```
Comment (nmussy) [inline packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Ideally, the `generation` and `windowsVersion` properties would be merged, but I didn't think it was worth the refactoring and breaking change.

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts [hunk 2]
```diff
@@ -267,9 +280,15 @@ export class EcsOptimizedAmi implements ec2.IMachineImage {
...
       if (props.generation === ec2.AmazonLinuxGeneration.AMAZON_LINUX && this.hwType !== AmiHardwareType.STANDARD) {
         throw new Error(`Amazon Linux does not support special hardware type. Use Amazon Linux 2 instead`);
+      } else if (props.windowsVersion) {
+        throw new Error('"windowsVersion" and Linux image "generation" cannot be both set');
       } else {
         this.generation = props.generation;
       }
+    } else if (props && props.windowsVersion) {
+      if (this.hwType !== AmiHardwareType.STANDARD) {
+        throw new Error('Windows Server does not support special hardware type. Use Amazon Linux 2 instead');
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Remove "use Amazon Linux 2 instead". I am assuming that if a user explicitly specified they wanted Windows, they should probably just not specify the HW type and not use Linux instead... :-)

File: packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts [hunk 3]
```diff
@@ -274,8 +274,73 @@ export = {
...
+    const cluster = new ecs.Cluster(stack, 'EcsCluster', { vpc });
+    cluster.addCapacity('WindowsAutoScalingGroup', {
+      instanceType: new ec2.InstanceType('t2.micro'),
+      machineImage: new ecs.EcsOptimizedAmi({
+        windowsVersion: ecs.WindowsOptimizedVersion.SERVER_2019,
+      }),
     });
 
+    // THEN
+    expect(stack).to(haveResource("AWS::AutoScaling::LaunchConfiguration", {
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: You should add an expectation on the parameter with the desired SSM path...

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts (hunk 2 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Done

File: packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts (hunk 3 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: Sorry, I don't understand your comment. I reused the existing HW AMI Type test: https://github.com/aws/aws-cdk/blob/47e81a63ff4b9d4326d56da2ac3b3fc04786cdd7/packages/%40aws-cdk/aws-ecs/test/test.ecs-cluster.ts#L252-L256

Comment (eladb) [reply packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: This just shows the ImageId is a reference to a synthesized CloudFormation parameter that is added to your template. You should write your expectation against the default value of the parameter which will contain the SSM parameter name that you want.

What I usually do is `expect(stack).toMatch({})` and then I can see which parts of the stack to assert against.

Comment (nmussy) [reply packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: Good call, I was missing an assignation, thanks. 

I have an issue with the matching now. I tried changing the `MatchStyle`, but it doesn't seem to detect when I purposefully change the `Default` matched value.

Should I use an integration test instead, to match the whole stack?

Comment (eladb) [reply packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: Yeah, I think there's a bug in the matcher (raise an issue?). What you could do in the meantime is as follows;

```ts
// you'll need an "App"
const app = new App();
const stack = new Stack(app, 'test');

// ...

// then
const assembly = app.synth();
const template = assembly.getStack(stack.stackName).template;
test.deepEqual(template.Parameters, { /* expectation here */ });
```

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Maybe we can @deprecate `generation` in favor of `amazonLinuxGeneration` (still support `generation` for backwards compat and in the next MV we will remove)

File: packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts (hunk 3 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: Alright, thanks a lot. Should I also update the existing Linux AMI test? https://github.com/aws/aws-cdk/blob/5c01a39ceb3f61832c743795ee68205c414ce53c/packages/%40aws-cdk/aws-ecs/test/test.ecs-cluster.ts#L253-L257

If so, should I do it in another PR?

Comment (eladb) [reply packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: Feel free to add to this PR

Comment (nmussy) [reply packages/@aws-cdk/aws-ecs/test/test.ecs-cluster.ts]: Done. Thanks again for the help!

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Thoughts? Maybe in a subsequent PR? At least update #3398 with this.

Comment (nmussy) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Sorry, I didn't see your previous comment. I can add it to this PR

Comment (nmussy) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Shouldn't we just replace it with an `image` union construct, to have the definition enforce the mutual exclusion?

Comment (eladb) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Yes!

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts [hunk 4]
```diff
@@ -291,7 +328,78 @@ export class EcsOptimizedAmi implements ec2.IMachineImage {
...
-      osType: ec2.OperatingSystemType.LINUX
+      osType: this.windowsVersion ? ec2.OperatingSystemType.WINDOWS : ec2.OperatingSystemType.LINUX
+    };
+  }
+}
+
+/**
+ * Construct a Linux or Windows machine image from the latest ECS Optimized AMI published in SSM
+ */
+export class EcsOptimizedAmiStatic implements ec2.IMachineImage {
```
Comment (eladb) [inline packages/@aws-cdk/aws-ecs/lib/cluster.ts]: Why is this called `Static`? If you are looking for a name that doesn't cause a breaking change, use `EcsOptimizedImage` or `EcsOptimizedMachineImage`

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts]: So what's up with this? Didn't we say this is gone?

**Pull request merged**